NebulaHost/PacketProcessors/Logistics/StationUIInitialSyncRequestProcessor.cs

```diff
@@ -0,0 +1,72 @@
+        {
+            StationComponent[] gStationPool = GameMain.data.galacticTransport.stationPool;
+            if(packet.stationGId < gStationPool.Length)
+            {
+                StationComponent stationComponent = gStationPool[packet.stationGId];
+                if(stationComponent != null)
+                {
+                    List<int> itemId = new List<int>();
+                    List<int> itemCountMax = new List<int>();
+                    List<int> localLogic = new List<int>();
+                    List<int> remoteLogic = new List<int>();
+                    for(int i = 0; i < stationComponent.storage.Length; i++)
+                    {
+                        itemId.Add(stationComponent.storage[i].itemId);
+                        itemCountMax.Add(stationComponent.storage[i].max);
+                        switch (stationComponent.storage[i].localLogic)
+                        {
+                            case (ELogisticStorage.None):
+                                localLogic.Add(0);
+                                break;
+                            case (ELogisticStorage.Supply):
+                                localLogic.Add(1);
+                                break;
+                            case (ELogisticStorage.Demand):
+                                localLogic.Add(2);
+                                break;
+                        }
+                        switch (stationComponent.storage[i].remoteLogic)
+                        {
+                            case (ELogisticStorage.None):
+                                remoteLogic.Add(0);
+                                break;
+                            case (ELogisticStorage.Supply):
+                                remoteLogic.Add(1);
+                                break;
+                            case (ELogisticStorage.Demand):
+                                remoteLogic.Add(2);
+                                break;
+                        }
+                    }
+                    StationUIInitialSync packet2 = new StationUIInitialSync(packet.stationGId,
+                                                                            stationComponent.tripRangeDrones,
+                                                                            stationComponent.tripRangeShips,
+                                                                            stationComponent.deliveryDrones,
+                                                                            stationComponent.deliveryShips,
+                                                                            stationComponent.warpEnableDist,
+                                                                            stationComponent.warperNecessary,
+                                                                            stationComponent.includeOrbitCollector,
+                                                                            itemId.ToArray(),
```

Comment:
If you pre-allocate the arrays above, you don't need to ToArray() here.

Reply:
is it bad to use `ToArray()` in this situation?

Reply:
ToArray() creates a copy of the collection.

```csharp
var list = new List<int>(); // this uses an int[8] internally
for (int i = 0; i < source.Length; ++i) {
   // If the list array is full, it doubles it, so you get another allocation for
   // int[16], int[32], int[64], etc.
   // you also need to copy hte values to the new array
   list.Add(source[i]); 
}
var result = list.ToArray(); // another `array[list.Count]` allocation + copy
```

if you just do
```csharp
// just one allocation and copy
var result = (int[])source.Clone();

// What clone does
var result = new int[source.Length];
Array.Copy(source, result, source.Length);

// Similarly you have
var result = Array.ConvertAll(source, x => x);
```

Reply:
Thanks for the detail example, that makes it really clear 😄